### ISSUE TITLE
Add option to specify dispatcher other than default for logger actors. 

### ISSUE THREAD

**Issue by Adebski** [akka/akka#20006]
Currently there are options to configure dispatchers for some parts of akka backend, e.g. remoting, clustering but (as far as I can see) there is no option to specify non-default dispatcher for actors that are part of logging infrastructure - `akka.loggers`. In some cases this could lead to 

* lost logs if all threads from `default-dispatcher` are blocked and someone decides to kill JVM before some of the threads are free again 
* significant delay in writing logs to desired output

Of course it would be nice if blocking tasks were delegated to some other specific pools instead of default one but because both remoting and clustering parts of akka use this logging infrastructure having an option to specify separate dispatcher could prove useful.

**Comment by patriknw** [comment]
There is typically only one `Logger` actor, so only one thread would be blocked by logging, or do I misunderstand your point?

**Comment by Adebski** [comment]
What I meant is that when all threads on default dispatcher are blocked (for whatever reason) the logs from remoting, clustering and all other parts of some system that may be working in the meantime won't be visible. In some situations instances of `LogEvent`s that are present in logging actor mailbox will be processed and written out (std, file - whatever is configured) with delay that depends on the time that all threads are blocked. 

Sometimes that could even result in lost log messages when someone decides to just restart the JVM because of blocked threads and contents of those `LogEvent`s will be lost. Of course when dealing with asynchronous logging one has to be aware of consequences but it would be nice to have an option to specify some other pool for such critical task.

**Comment by guidomedina** [comment]
There are few other things that you will probably need to contribute such functionality and make it nicely:
- a default pine dispatcher (as it is only one actor as @patriknw mentioned) defined in the reference.conf which the user can change to some other dispatcher if he wants to.
- make Akka aware of such dispatcher so that it is shutdown as last; same as the actor? Akka shutdown lifecycle awareness.
- a MPSC mailbox configuration key for such actor, maybe you want to pass a non-blocking bounded queue and ignore messages after some threshold; there are only two mailbox semantics at the moment so I think there shouldn't be any enforcement for its mailbox.

_These are the things that IMHO I would like to see in such functionality if and when added._

**Comment by patriknw** [comment]
@Adebski you should not run blocking things on the default dispatcher that risk to block all its threads. Then you will see much worse things than lack of logging.

I agree that it would be nice to be able to configure a dispatcher for the logging actor(s), but my point is that it's not the root cause of your problems.

**Comment by ktoso** [comment]
I think it would be a nice-to-have (will mark as such), however doesn't really solve a core issue (backing logging by an high perf async appender is something I'd hope people do on production, and providing such is entire other OSS projects already doing that :-))

Let's keep as a nice-to-have, if someone wants to give it a shot.

**Comment by guidomedina** [comment]
Two rhetorical questions:

- is it a nice to have? yes, but ...
- is it the current akka logging system half completed in that it runs under the default dispatcher and nothing else can be changed? yes
- can it be annoying for actual production systems? yes, think of the fact that you are offering an asynchronous logging (an actor) that is can potentially do more harm than help, then ask the question again: is it a nice to have?

So, it is a bit more than a nice to have.

**Comment by guidomedina** [comment]
Also, other loggers offer such functionality indeed...with dedicated thread, some of them by adding yet another dependency.

**Comment by ktoso** [comment]
My opinion is, and remains, that it is not our highest priority, as we have plenty of actual urgent things to work on in other areas of the project. For prod I would recommend using an async appender in any case, and we won't go into the rabbit hole of attempting to beat the top async appenders right now (would be fun, but not a priority) :-)

Also note @guidomedina that this is a logging bridge, but full blown logging framework – we highly recommend using the akka-slf4j module, with for example `log4j2` as that is exactly what that project is about – just logging. The LoggingAdapter is in akka-actor in order to not have any dependencies in akka-actor other than the JDK.

The day only has 24hours, and this is not the top priority for the our team – if you'd like to contribute that's _exactly_ what those `nice-to-have (low-prio)` marker tag is for, for marking such areas – thanks in advance!

**Comment by patriknw** [comment]
nice-to-have also means that it is a great thing for the community to help out with. This should not be difficult, if not overengineered, so we can mark it easy-to-contribute

**Comment by guidomedina** [comment]
Working on this one, got the dispatcher part done, now trying to figure it out if is worth the effort of allowing mailbox but the user will have to be aware of the drain method and `LoggerMailboxSemantics`

**Comment by guidomedina** [comment]
you know what, not going to over-complicate it, let me PR what I have and then we can decide (or someone else more a Scala expert can improve it)

**Comment by guidomedina** [comment]
@ktoso branch name for this?

**Comment by ktoso** [comment]
Thanks, let's see!
Branch name is not important (commit message is see CONTRIBUTING.md - should refer this issue in message)

**Comment by guidomedina** [comment]
Check commit before I PR please: https://github.com/guidomedina/akka/commit/325200e90794d16284649700726088b343af9878

**Comment by ktoso** [comment]
Please PR it, makes it much easier to find and comment :)
